base/shared/src/main/scala/scalaz/data/io.scala

```diff
@@ -0,0 +1,83 @@
+package scalaz
+package data
+
+import scalaz.zio.Errors.NothingRaced
+import scalaz.zio.{ Fiber, IO }
+import scalaz.tc.Foldable
+
+import scala.{ AnyVal, Either, Nothing, Option, Unit }
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+import scala.inline
+
+object io {
+
+  private final def _bracket0[E, A, B](
+    acquire: IO[E, A]
+  )(release: (A, Maybe[E \/ B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] = {
+    val release1: (A, Option[Either[E, B]]) => IO[Nothing, Unit] =
+      (a: A, o: Option[Either[E, B]]) => release(a, Maybe.fromOption(o.map(Disjunction.fromEither[E, B])))
+    IO.bracket0[E, A, B](acquire)(release1)(use)
+  }
+
+  implicit class IOObjOps(private val ioObj: IO.type) extends AnyVal {
+
+    final def absolve[E, A](v: IO[E, E \/ A]): IO[E, A] =
+      v.flatMap(fromDisjunction)
+
+    final def fromDisjunction[E, A](v: E \/ A): IO[E, A] =
+      v.fold(IO.fail, IO.now)
+
+    @inline
+    final def bracket0[E, A, B](
+      acquire: IO[E, A]
+    )(release: (A, Maybe[E \/ B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
+      _bracket0[E, A, B](acquire)(release)(use)
+
+    final def require[E, A](error: E): IO[E, Maybe[A]] => IO[E, A] =
+      (io: IO[E, Maybe[A]]) =>
+        io.flatMap { maybe =>
+          Maybe.fold[A, IO[E, A]](maybe)(IO.now[A], IO.fail[E](error))
+      }
+
+    // TODO replace CanBuildFrom with Unfoldable
+    final def forkAll[E, A, F[_]](
+      as: F[IO[E, A]]
+    )(implicit f: Foldable[F], cbf: CanBuildFrom[F[IO[E, A]], A, F[A]]): IO[E, Fiber[E, F[A]]] =
+      f.foldRight(as, IO.point[Fiber[E, mutable.Builder[A, F[A]]]](Fiber.point(cbf(as)))) {
+          case (a, as) => as.par(a.fork).map { case (as, a) => as.zipWith(a)(_ += _) }
+        }
+        .map(as => as.zipWith(Fiber.point(())) { case (as, _) => as.result })
+
+    final def raceAll[E, A, F[_]](t: F[IO[E, A]])(implicit f: Foldable[F]): IO[E, A] =
+      f.foldLeft[IO[E, A], IO[E, A]](t, IO.terminate(NothingRaced))(_ race _)
+
+    final def reduceAll[E, A, F[_]](a: IO[E, A], as: F[IO[E, A]])(f: (A, A) => A)(implicit fd: Foldable[F]): IO[E, A] =
+      fd.foldLeft(as, a) { (l, r) =>
+        l.par(r).map(f.tupled)
+      }
+
+    final def mergeAll[E, A, B, F[_]](in: F[IO[E, A]])(zero: B, f: (B, A) => B)(implicit fd: Foldable[F]): IO[E, B] =
+      fd.foldLeft[IO[E, A], IO[E, B]](in, IO.point[B](zero))((acc, a) => acc.par(a).map(f.tupled))
+  }
+
+  implicit class IOOps[E, A](private val io: IO[E, A]) extends AnyVal {
+    final def attempt: IO[Nothing, E \/ A] =
```

Comment:
Not sure Scala will find this one, unfortunately.

Reply:
Ouch.. just got bitten by this while writing tests..

```scala
test("bracket0") { () =>
  def release[A, E, B]: (A, Maybe[E \/ B]) => IO[Nothing, Unit] = (a, meb) => IO.point(())
  def use[A, E]: A => IO[E, A] = a => IO.point(a)

  val res: Int = unsafeRun(IO.point[Int](1).bracket0(release[Int, Nothing, Int])(use[Int, Nothing]))
  assertEqual(1, 1) // TODO
}
```

This is returning

> [error] type mismatch;
[error]  found   : (Int, scalaz.data.Maybe[Nothing \/ Int]) => scalaz.zio.IO[Nothing,Unit]
[error]     (which expands to)  (Int, scalaz.data.Maybe.Maybe[scalaz.data.Disjunction[Nothing,Int]]) => scalaz.zio.IO[Nothing,Unit]
[error]  required: (Int, Option[Either[?,?]]) => scalaz.zio.IO[Nothing,Unit]
[error]           val res: Int = unsafeRun(IO.point[Int](1).bracket0(release[Int, Nothing, Int])(use[Int, Nothing]))
    
Any ideas?

Reply:
In order for Scala to use the implicit, there cannot exist a method of the same name on the target with the same erased type. So the options are to change the type, so Scala can find it, or to change the name.

Reply:
I guess we should find new names then..
I wouldn't touch the types.
Any conventions here, or I'm free to propose new ones?
Thanks @jdegoes , I'm learning a lot.